File: 02_algorithm/programmers/Level2/(2020-KAKAO-상반기-인턴십)수식 최대화.py
import itertools
import copy

def calc(num_1, num_2, oper):
    if oper == '+':
        return num_1 + num_2
    elif oper == '-':
        return num_1 - num_2
    else:
        return num_1 * num_2


def temp_calc(numbers, operators, operator_rank):
    for oper in operator_rank:
        while oper in operators:
            oper_idx = operators.index(oper)
            temp_value = calc(numbers[oper_idx], numbers[oper_idx + 1], oper)
            numbers[oper_idx + 1] = temp_value
            numbers.pop(oper_idx)
            operators.pop(oper_idx)
    return abs(numbers[0])


def solution(expression):
    answer = 0
    numbers, operators = [], []
    temp_num = ''
    for express in expression:
        if express not in ['+', '-', '*']:
            temp_num += express
        else:
            numbers.append(int(temp_num))
            operators.append(express)
            temp_num = ''
    numbers.append(int(temp_num))
    for comb in itertools.permutations(set(operators), len(set(operators))):
        original_nums, original_opers = copy.deepcopy(numbers), copy.deepcopy(operators)
        answer = max(answer, temp_calc(numbers, operators, comb))
        numbers, operators = original_nums, original_opers
    return answer


print(solution('100-200*300-500+20'))
print(solution('50*6-3*2'))

File: 06_js/04_algorithm/programmers/Level2/(2020-KAKAO-상반기-인턴십)수식 최대화.js
function calc(num1, num2, oper) {
  if (oper === '+') {
    return num1 + num2;
  } else if (oper === '-') {
    return num1 - num2;
  } else {
    return num1 * num2;
  }
}


function calcFormula(numbers, operators, priority) {
  for (const operator of priority) {
    const operatorIndexes = operators
      .map((oper, idx) => oper === operator && idx)
      .filter(val => val === 0 || val !== false);
    const popIndex = [];
    for (const operatorIndex of operatorIndexes) {
      const tempCalcValue = calc(numbers[operatorIndex], numbers[operatorIndex + 1], operator);
      numbers[operatorIndex + 1] = tempCalcValue;
      numbers[operatorIndex] = 0;
      popIndex.push(operatorIndex);
    }
    for (const idx of popIndex.reverse()) {
      numbers.splice(idx, 1);
      operators.splice(idx, 1);
    }
  }
  return Math.abs(numbers[0]);
}


function getOperatorComb(operatorsSet) {
  const operators = Array.from(operatorsSet);
  const operatorPriorities = [];
  const getComb = (operators, comb) => {
    if (comb.length === operators.length) {
      operatorPriorities.push([...comb]);
      return
    }
    for (const operator of operators) {
      if (!comb.includes(operator)) {
        comb.push(operator);
        getComb(operators, comb);
        comb.pop();
      }
    }
  }
  getComb(operators, []);
  return operatorPriorities;
}


function solution(expression) {
  let answer = 0;
  // (1) 숫자, 기호 분리
  const numbers = [];
  const operators = [];
  let tempNumber = '';
  for (const express of expression) {
    if (isNaN(express)) {
      operators.push(express);
      numbers.push(Number(tempNumber));
      tempNumber = '';
      continue;
    }
    tempNumber += express;
  }
  numbers.push(Number(tempNumber));
  
  // (2) 가능한 연산자 우선순위 모든 경우의 수 구하기
  const operatorsSet = new Set(operators);
  const operatorPriorities = getOperatorComb(operatorsSet);
  
  // (3) 연산자 우선순위 모든 경우에 대한 계산 결과 구해서 절댓값 적용된 최댓값 구하기
  for (const priority of operatorPriorities) {
    answer = Math.max(answer, calcFormula([...numbers], [...operators], priority));
  }
  return answer;
}

console.log(solution('100-200*300-500+20'));
console.log(solution('50*6-3*2'));